### PR TITLE
test(robot): add test backup listing for large volumes and many backups

### PR DESCRIPTION
**Pull request by roger-ryao**
#### Which issue(s) this PR fixes:
longhorn/longhorn#8355

#### What this PR does / why we need it:
Implement manual test case [Test backup listing S3/NFS](https://longhorn.github.io/longhorn-tests/manual/pre-release/stress/backup-listing/) in robot.

#### Additional documentation or context
- https://longhorn.github.io/longhorn-tests/manual/pre-release/stress/backup-listing/
- https://github.com/longhorn/longhorn/issues/8355
- PR test log : 
  - ~~[log_20240617_case2.tar.gz](https://github.com/user-attachments/files/15860633/log_20240617_case2.tar.gz)~~
  - ~~[log_20240823.tar.gz](https://github.com/user-attachments/files/16720876/log_20240823.tar.gz)~~
  - [log_20240926.tar.gz](https://github.com/user-attachments/files/17143932/log_20240926.tar.gz)

  - Here's how to execute test case: ` ./run.sh -t "Backup listing of volume bigger than 200 Gi" -v RETRY_COUNT:5400
`

### CODE REVIEW

File: e2e/tests/test_cases/test_backup_listing.robot [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    Create volume ${volume_id} with ${size} GB and 3 replicas
+    Create persistentvolume for volume ${volume_id}
+    Create persistentvolumeclaim for volume ${volume_id}
+    Create pod ${pod_id} using volume ${volume_id}
+    Wait for pod ${pod_id} running
+
+Create backup ${backup_id} for volume ${volume_id} in ${timeout_in_min} min
+    ${timeout_in_sec}=    Evaluate    ${timeout_in_min} * 60
+    ${volume_name}=    generate_name_with_suffix    volume    ${volume_id}
+    create_backup    ${volume_name}    ${backup_id}    ${timeout_in_sec}
```
Comment (yangchiu) [inline e2e/tests/test_cases/test_backup_listing.robot:96]: Why do we need to add this timeout argument?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/tests/test_cases/test_backup_listing.robot (hunk 1 above)
Comment (yangchiu) [reply e2e/tests/test_cases/test_backup_listing.robot:96]: Why not pass the customized `RETRY_COUNT` parameter:
```
-v RETRY_COUNT:259200
```
That's why we make it a variable.

Comment (roger-ryao) [reply e2e/tests/test_cases/test_backup_listing.robot:96]: Good suggestion. When I previously executed the test case, I did not use the `RETRY_COUNT` parameter. I will revert that change and test again.

**Comment by yangchiu** [comment]
@roger-ryao Please let the team know if this PR is ready for review. Thank you!

**Comment by roger-ryao** [comment]
> @roger-ryao Please let the team know if this PR is ready for review. Thank you!

Hi @yangchiu 
    The PR is ready for review.

File: e2e/libs/backup/rest.py [hunk 2]
```diff
@@ -17,6 +17,7 @@ def __init__(self):
 
     def create(self, volume_name, backup_id):
         # create snapshot
+        logging(f"volume is {volume_name} bckupid is {backup_id}")
```
Comment (yangchiu) [inline e2e/libs/backup/rest.py:20]: It would be better to have a more meaningful logging.

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/keywords/snapshot.resource [hunk 3]
```diff
@@ -13,6 +13,13 @@ Delete snapshot ${snapshot_id} of volume ${volume_id}
     ${volume_name} =    generate_name_with_suffix    volume    ${volume_id}
     delete_snapshot    ${volume_name}    ${snapshot_id}
 
+Delete snapshot ${snapshot_id} of ${workload_kind} ${workload_id} volume
+    ${snapshot_id}=  Convert To String  ${snapshot_id}
+    ${workload_name}=   generate_name_with_suffix    ${workload_kind}    ${workload_id}
+    ${volume_name}=    get_workload_volume_name    ${workload_name}
+    delete_snapshot    ${volume_name}    ${snapshot_id}
+    purge_snapshot    ${volume_name}
```
[the remark this answers is not in the document]
Comment (roger-ryao) [reply e2e/keywords/snapshot.resource:21]: "Maximum snapshot count for a volume: The value should be between 2 and 250. If purge is not called here, [creating the 251st snapshot will fail](https://github.com/longhorn/longhorn-tests/pull/1940/files#diff-efad252d110ad71c3aa1a49d32bf7b7d7b1c1b77de8ff754f3307e8406f6f71bR83-R90).

File: e2e/keywords/volume.resource [hunk 4]
```diff
@@ -225,10 +232,12 @@ Activate DR volume ${volume_id}
     activate_dr_volume    ${volume_name}
 
 Create persistentvolume for volume ${volume_id}
+    ${volume_id}=  Convert To String  ${volume_id}
```
[the remark this answers is not in the document]
Comment (roger-ryao) [reply e2e/keywords/volume.resource:235]: No, `Convert To String` on [Line 235](https://github.com/longhorn/longhorn-tests/pull/1940/files#diff-6422ed87ee022691c861ebd060f6651dd79d25bef034662a78dc201f40a8838bR235) and [Line 240](https://github.com/longhorn/longhorn-tests/pull/1940/files#diff-6422ed87ee022691c861ebd060f6651dd79d25bef034662a78dc201f40a8838bR240) are an unnecessary and redundant steps. I will remove it.

File: e2e/keywords/snapshot.resource [hunk 5]
```diff
@@ -13,6 +13,13 @@ Delete snapshot ${snapshot_id} of volume ${volume_id}
     ${volume_name} =    generate_name_with_suffix    volume    ${volume_id}
     delete_snapshot    ${volume_name}    ${snapshot_id}
 
+Delete snapshot ${snapshot_id} of ${workload_kind} ${workload_id} volume
+    ${snapshot_id}=  Convert To String  ${snapshot_id}
```
[the remark this answers is not in the document]
Comment (roger-ryao) [reply e2e/keywords/snapshot.resource:17]: Yes, If we don't add `Convert To String` , the keyword `Perform backup 1001 times for deployment 0 volume` will fail.
Screenshot_20240826_134217

File: e2e/keywords/backup.resource [hunk 6]
```diff
@@ -9,15 +9,33 @@ Create backup ${backup_id} for volume ${volume_id}
...
+    ${volume_name} =    get_workload_volume_name    ${workload_name}
+    verify_no_error    ${volume_name}
+
 Verify backup list contains backup ${backup_id} of volume ${volume_id}
     ${volume_name} =    generate_name_with_suffix    volume    ${volume_id}
     ${backup} =    get_backup_name    ${backup_id}    ${volume_name}
     Should Not Be Equal    ${backup}    ${None}
 
+Verify backup list contains backup ${backup_id} of ${workload_kind} ${workload_id} volume
+    ${backup_id}=  Convert To String  ${backup_id}
```
[the remark this answers is not in the document]
Comment (roger-ryao) [reply e2e/keywords/backup.resource:33]: Yes, If we don't  add `Convert To String` , the keyword `Perform backup 1001 times for deployment 0 volume` will fail.
Screenshot_20240826_134523

File: e2e/keywords/workload.resource [hunk 7]
```diff
@@ -27,6 +27,15 @@ Keep writing data to pod of ${workload_kind} ${workload_id}
     ${workload_name} =   generate_name_with_suffix    ${workload_kind}    ${workload_id}
     keep_writing_workload_pod_data    ${workload_name}
 
+Write ${size} MB data to file ${file_name} in pod ${pod_id}
+    ${pod_name} =    generate_name_with_suffix    pod    ${pod_id}
+    write_workload_pod_random_data    ${pod_name}    ${size}    ${file_name}
+
+Write ${size} GB data to file ${file_name} in pod ${pod_id}
+    ${file_name} =  Convert To String  ${file_name}
```
[the remark this answers is not in the document]
Comment (roger-ryao) [reply e2e/keywords/workload.resource:35]: No, it is an unnecessary and redundant steps. I will remove it.

File: e2e/keywords/snapshot.resource (hunk 3 above)
[the remark this answers is not in the document]
Comment (yangchiu) [reply e2e/keywords/snapshot.resource:21]: Should it be moved to `Perform backup 1001 times for deployment 0 volume` to explicitly call `purge` every 250 backups?

**Review by yangchiu — CHANGES_REQUESTED** [review]

**Comment by roger-ryao** [comment]
The PR is ready for review.

File: e2e/tests/negative/test_backup_listing.robot [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    And Write data to file in deployment 0
+    Then Perform backup 1001 times for deployment 0 volume
+    Then Create volume 1 from deployment 0 volume random backup
+    And Create deployment 1 with volume 1
+    Then Get deployment 1 volume data in file data
+    And Volume 1 data should same as deployment 0 volume
+    Then Create pod 2 mount 250 GB volume 2
+    And Write 210 GB data in pod 2
+    Then Volume 2 backup 0 should be able to create
+    Then Delete pod 2 and volume 2
```
Comment (yangchiu) [inline e2e/tests/negative/test_backup_listing.robot:172]: Is the deletion necessary here? Can we handle pod and volume deletion within the teardown cleanup function instead?

File: e2e/libs/workload/workload.py [hunk 9]
```diff
@@ -102,6 +102,33 @@ def write_pod_random_data(pod_name, size_in_mb, file_name,
...
+    retry_count, retry_interval = get_retry_count_and_interval()
+
+    for _ in range(retry_count):
+        try:
+            data_path = f"{data_directory}/{file_name}"
+            api = client.CoreV1Api()
+            write_data_cmd = [
+                '/bin/sh',
+                '-c',
+                f"fallocate -l {size_in_gb}G {data_path};\
```
Comment (yangchiu) [inline e2e/libs/workload/workload.py:119]: How long will the test case take after adopting `fallocate`?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/tests/negative/test_backup_listing.robot [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+Volume ${volume_id} backup ${backup_id} should be able to create
+    Create backup ${backup_id} for volume ${volume_id}
+    Verify backup list contains no error for volume ${volume_id}
+    Verify backup list contains backup ${backup_id} of volume ${volume_id}
+
+Write data to file in deployment 0
+    Write 100 MB data to file data in deployment 0
+
+Write 210 GB data in pod 2
```
Comment (yangchiu) [inline e2e/tests/negative/test_backup_listing.robot:112]: I don’t think it’s necessary to wrap it in another keyword. Using `Write 210 GB large data to file 0 in pod 2` directly in the test steps should be sufficient. WDYT?

**Review by yangchiu — CHANGES_REQUESTED** [review]

File: e2e/libs/workload/workload.py (hunk 9 above)
Comment (roger-ryao) [reply e2e/libs/workload/workload.py:119]: Reduced from approximately 13 hours to six hours.
Screenshot_20241013_215127

File: e2e/tests/negative/test_backup_listing.robot (hunk 10 above)
Comment (roger-ryao) [reply e2e/tests/negative/test_backup_listing.robot:112]: I agree with your suggestion, and I will make changes to this part.

File: e2e/tests/negative/test_backup_listing.robot (hunk 8 above)
Comment (roger-ryao) [reply e2e/tests/negative/test_backup_listing.robot:172]: In this case, we need to create a 250GB `Volume 2` and then create `Volume 3` by `restoring from the backup of Volume 2`. In the test environment where I develop test cases, each worker node has a 1TB hard drive. However, I haven't confirmed the minimum resource requirements for this case. If we want to use a shared cluster for each case, I believe interspersing deletions in the case store can help us avoid using excessively large disk resources.

**Comment by roger-ryao** [comment]
Hi @longhorn/qa 
     The PR is ready for review.

**Comment by roger-ryao** [comment]
> This pull request is now in conflict. Could you fix it @roger-ryao? 🙏

Done

**Review by yangchiu — APPROVED** [review]
LGTM

**Pull request merged**